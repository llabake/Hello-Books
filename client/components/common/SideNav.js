import React, { Component } from 'react';
import { Link, NavLink } from 'react-router-dom';
import { connect } from 'react-redux'
import { logout } from '../../actions/userAction'

import SearchBar from './SearchBar';

/**
 *
 * @param {*} { user, authenticated, handleLogout }
 * 
 * @returns {JSX} side nav
 */
const SideNav = ({ user, authenticated, handleLogout }) => {
  /**
   *
   * @param {*} normalUser
   * 
   * @returns {Object} auhtnticated user nav bar
   * 
   * @memberof SideNav
   */
  const renderAuthUserSideNavigation = (normalUser) => {
    return (
      <li class="no-padding">
        <ul class="collapsible collapsible-accordion">
          <li>
            <a class="collapsible-header">{normalUser.username}
              <i class="material-icons">account_circle</i>
              <i class="material-icons right">arrow_drop_down</i>
            </a>
            <div class="collapsible-body">
              <ul>
                <li><Link to="/allbooks">All Books</Link></li>
                <li><Link to="/favorite">Favorite Books</Link></li>
                <li><Link to="/profile">Profile</Link></li>
                <li className="divider"></li>
                <li>
                  <NavLink to=""
                    onClick={handleLogout}>
                    <i className="material-icons ">lock</i>Log Out
                  </NavLink>
                </li>
              </ul>
            </div>
          </li>
        </ul>
      </li>
    )
  }

  /**
   *
   * @param {*} adminUser
   * 
   * @returns {Object} admin navigation side bar
   * 
   * @memberof SideNav
   */
  const renderAdminSideNavigation = (adminUser) => {
    if (user.role === 'admin') {
      return (
        <li class="no-padding">
          <ul class="collapsible collapsible-accordion">
            <li>
              <a class="collapsible-header">{adminUser.username}
                <i class="material-icons">account_circle</i>
                <i class="material-icons right">arrow_drop_down</i>
              </a>
              <div class="collapsible-body">
                <ul>
                  <li><Link to="/admindashboard">Admin Dashboard</Link></li>
                  <li><Link to="/favorite">Favorite Books</Link></li>
                  <li><Link to="/profile">Profile</Link></li>
                  <li><Link to="/addbook">Add Book</Link></li>
                  <li><Link to="/allbooks">All Books</Link></li>
                  <li className="divider"></li>
                  <li>
                    <Link to="" onClick={handleLogout}>
                      <i className="material-icons ">lock</i>Log Out
                  </Link>
                  </li>
                </ul>
              </div>
            </li>
          </ul>
        </li>
      )
    }
  }

  const renderSideNav = (authUser) => authUser.role === 'admin'
    ? renderAdminSideNavigation(user)
    : renderAuthUserSideNavigation(user)

  return (
    <ul id="mobile-demo" class="side-nav">
      <li>
        <SearchBar />
      </li>
      {authenticated && user
        ? renderSideNav(user)
        :
        <li class="no-padding">
          <ul class="collapsible collapsible-accordion">
            <li>
              <a class="collapsible-header">Hi, User
            <i class="material-icons">account_circle</i>
                <i class="material-icons right">arrow_drop_down</i>
              </a>
              <div class="collapsible-body">
                <ul>
                  <li><Link to="/signup">Sign Up</Link></li>
                  <li><Link to="/signin">Sign In</Link></li>
                </ul>
              </div>
            </li>
          </ul>
        </li>
      }
    </ul>
  )
}

export default SideNav;