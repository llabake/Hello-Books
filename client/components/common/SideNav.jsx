import React from 'react';
import PropTypes from 'prop-types';
import { Link, NavLink } from 'react-router-dom';
import { connect } from 'react-redux'

import SearchBar from './SearchBar';

/**
 *
 * @param {*} { user, authenticated, handleLogout }
 * 
 * @returns {JSX} side nav
 */
const SideNav = ({ user, authenticated, handleLogout }) => {
  /**
   *
   * @param {*} normalUser
   * 
   * @returns {Object} auhtnticated user nav bar
   * 
   * @memberof SideNav
   */
  const renderAuthUserSideNavigation = (normalUser) => {
    return (
      <li className="no-padding">
        <ul className="collapsible collapsible-accordion">
          <li>
            <a className="collapsible-header">{normalUser.username}
              <i className="material-icons">account_circle</i>
              <i className="material-icons right">arrow_drop_down</i>
            </a>
            <div className="collapsible-body">
              <ul>
                <li><Link to="/allbooks">All Books</Link></li>
                <li><Link to="/favorite">Favorite Books</Link></li>
                <li><Link to="/profile">Profile</Link></li>
                <li className="divider"></li>
                <li>
                  <NavLink to="" onClick={handleLogout}>
                    <i className="material-icons ">lock</i>Log Out
                  </NavLink>
                </li>
              </ul>
            </div>
          </li>
        </ul>
      </li>
    )
  }

  /**
   *
   * @param {*} adminUser
   * 
   * @returns {Object} admin navigation side bar
   * 
   * @memberof SideNav
   */
  const renderAdminSideNavigation = (adminUser) => {
    if (user.role === 'admin') {
      return (
        <li className="no-padding">
          <ul className="collapsible collapsible-accordion">
            <li>
              <a className="collapsible-header">{adminUser.username}
                <i className="material-icons">account_circle</i>
                <i className="material-icons right">arrow_drop_down</i>
              </a>
              <div className="collapsible-body">
                <ul>
                  <li><Link to="/admindashboard">Admin Dashboard</Link></li>
                  <li><Link to="/favorite">Favorite Books</Link></li>
                  <li><Link to="/profile">Profile</Link></li>
                  <li><Link to="/addbook">Add Book</Link></li>
                  <li><Link to="/allbooks">All Books</Link></li>
                  <li className="divider"></li>
                  <li>
                    <Link to="" onClick={handleLogout}>
                      <i className="material-icons ">lock</i>Log Out
                  </Link>
                  </li>
                </ul>
              </div>
            </li>
          </ul>
        </li>
      )
    }
  }

  const renderSideNav = (authUser) => authUser.role === 'admin'
    ? renderAdminSideNavigation(user)
    : renderAuthUserSideNavigation(user)

  return (
    <ul id="mobile-demo" className="side-nav">
      <li>
        <SearchBar />
      </li>
      {authenticated && user
        ? renderSideNav(user)
        :
        <li className="no-padding">
          <ul className="collapsible collapsible-accordion">
            <li>
              <a className="collapsible-header">Hi, User
            <i className="material-icons">account_circle</i>
                <i className="material-icons right">arrow_drop_down</i>
              </a>
              <div className="collapsible-body">
                <ul>
                  <li><Link to="/signup">Sign Up</Link></li>
                  <li><Link to="/signin">Sign In</Link></li>
                </ul>
              </div>
            </li>
          </ul>
        </li>
      }
    </ul>
  )
}

SideNav.propTypes = {
  user: PropTypes.shape({
    username: PropTypes.string,
    email: PropTypes.string,
    role: PropTypes.string,
    id: PropTypes.number,
    active: PropTypes.bool,
  }).isRequired,
  authenticated: PropTypes.bool,
  handleLogout: PropTypes.func
}

export default SideNav;