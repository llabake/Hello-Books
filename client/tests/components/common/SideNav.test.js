import React from 'react'
import { NavLink } from 'react-router-dom'
import SideNav from '../../../components/common/SideNav'
import userData from '../../mocks/userData'

test('The SideNav component mounts correctly if all props it needs are passed into it.', () => {
  const props = {
    authenticated: false,
    user: {},
    errors: {},
    logout: jest.fn()
  }

  const wrapper = shallow(<SideNav {...props} />)
  expect(wrapper).toMatchSnapshot()
})

test('the user details are displayed when the user is logged in', () => {
  const props = {
    authenticated: true,
    user: userData.user,
    errors: {}
  }
  const wrapper = shallow(<SideNav {...props}  user={props.user} />)
  const usernameDropdown = wrapper.find('.collapsible-header')
  expect(usernameDropdown.text()).toMatch(`${props.user.username}account_circlearrow_drop_down`)
})

test('the admin details are displayed when the user is logged in as an admin', () => {
  const props = {
    authenticated: true,
    user: userData.admin,
    errors: {}
  }
  const wrapper = shallow(<SideNav {...props} user={props.user} role = {props.user.role}/>)
  const usernameDropdown = wrapper.find('.collapsible-header')
  expect(usernameDropdown.text()).toMatch(`${props.user.username}account_circlearrow_drop_down`)
});

