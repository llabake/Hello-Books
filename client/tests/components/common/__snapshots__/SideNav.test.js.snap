// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SideNav component mounts correctly if all props it needs are passed into it. 1`] = `
<ul
  className="side-nav show-on-medium-and-down"
  id="mobile-demo"
>
  <li>
    <Connect(SearchBar) />
  </li>
  <li
    className="no-padding"
  >
    <ul
      className="collapsible collapsible-accordion"
    >
      <li>
        <a
          className="collapsible-header"
        >
          Hi, User
          <i
            className="material-icons"
          >
            account_circle
          </i>
          <i
            className="material-icons right"
          >
            arrow_drop_down
          </i>
        </a>
        <div
          className="collapsible-body"
        >
          <ul>
            <li>
              <Link
                replace={false}
                to="/signup"
              >
                Sign Up
              </Link>
            </li>
            <li>
              <Link
                replace={false}
                to="/signin"
              >
                Sign In
              </Link>
            </li>
          </ul>
        </div>
      </li>
    </ul>
  </li>
</ul>
`;
